registry/storage/driver/middleware/alicdn/middleware.go

```diff
@@ -0,0 +1,110 @@
+package middleware
+
+import (
+	"fmt"
+	"hash/fnv"
+
+	"github.com/docker/distribution/context"
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+	storagemiddleware "github.com/docker/distribution/registry/storage/driver/middleware"
+	"net/url"
+	"time"
+)
+
+type aliCDNStorageMiddleware struct {
+	storagedriver.StorageDriver
+	scheme string
+	host   string
+}
+
+func newAliCDNStorageMiddleware(storageDriver storagedriver.StorageDriver, options map[string]interface{}) (storagedriver.StorageDriver, error) {
+	o, ok := options["baseurl"]
+	if !ok {
+		return nil, fmt.Errorf("no baseurl provided")
+	}
+	b, ok := o.(string)
+	if !ok {
+		return nil, fmt.Errorf("baseurl must be a string")
+	}
+	u, err := url.Parse(b)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse alicdn baseurl: %s", b)
+	}
+	if u.Scheme == "" {
+		return nil, fmt.Errorf("no scheme specified for alicdn baseurl")
+	}
+	if u.Host == "" {
+		return nil, fmt.Errorf("no host specified for alicdn baseurl")
+	}
+	return &aliCDNStorageMiddleware{
+		StorageDriver: storageDriver,
+		scheme:        u.Scheme,
+		host:          u.Host,
+	}, nil
+}
+
+func (cdn *aliCDNStorageMiddleware) h(data string) int {
+	h := fnv.New32a()
+	h.Write([]byte(data))
+	return int(h.Sum32())
+}
+
+func (cdn *aliCDNStorageMiddleware) expireOf(now time.Time, path string) time.Time {
+
+	// Every file has it's own expire point based on hash of the path,
+	// such as 19 minutes past the hour, 25 minutes and 12 seconds past the hour, to avoid heavy traffic
+	// back to oss if all file on cdn expire at the same time.
+
+	// If the calculated expire time is too close(less than 5 minutes), return expire point at next hour.
+
+	// To balance the security and cost, one hour as max expire time is considerable.
+
+	secondsInWallClock := cdn.h(path) % 3600
+	now = now.Truncate(time.Second)
+	secondsOfHour := now.Minute()*60 + now.Second()
+	insurance := 5 * 60
+
+	var expire time.Time
+	expire = now.Add(time.Duration(secondsInWallClock-secondsOfHour) * time.Second)
+	if secondsInWallClock-secondsOfHour < insurance {
```

Comment:
Reason I asked about documentation is because this algorithm almost seems very specific. Instead of just using the hash as a seed for randomizing the expiration time or something much simpler.

Based on the math above...
Max value of `secondsOfHour` would be 3599 and the min value of `secondsInWallClock` would be 0, allowing for a minimum offset of `-3599`. When anything less than `-3299` is calculated, the expiration will still be within 5 minutes. 

Instead of just fixing the math here to add expiration, would like to see external documentation for this algorithm or something simpler. If the goal is simply to spread out the expirations throughout the hour, that is fair and let's see something more straightforward.